src/Database/Connection.php

```diff
@@ -313,8 +367,10 @@ public function isConnected(): bool
      */
     public function prepare($query): StatementInterface
     {
-        return $this->getDisconnectRetry()->run(function () use ($query) {
-            $statement = $this->_driver->prepare($query);
+        $role = $query instanceof Query ? $query->getConnectionRole() : self::ROLE_WRITE;
+
+        return $this->getDisconnectRetry()->run(function () use ($query, $role) {
```

Comment:
If `getDisconnectRetry()` took a role name we could skip having to connect both connections.